src/main/java/org/tugraz/sysds/runtime/data/TensorBlock.java

```diff
@@ -92,15 +271,344 @@ public boolean isEmpty() {
+		if (_heterogeneous)
+			_dataTensor.set(v);
+		else
+			_basicTensor.set(v);
+	}
 
-	public abstract void set(int r, int c, double v);
+	public void set(MatrixBlock other) {
+		if (_heterogeneous)
+			throw new DMLRuntimeException("TensorBlock.set(MatrixBlock) is not yet implemented for heterogeneous tensors");
+		else
+			_basicTensor.set(other);
+	}
+
+	public void set(int[] ix, Object v) {
+		if (_heterogeneous)
+			_dataTensor.set(ix, v);
+		else
+			_basicTensor.set(ix, v);
+	}
+
+	public void set(int r, int c, double v) {
+		if (_heterogeneous)
+			_dataTensor.set(r, c, v);
+		else
+			_basicTensor.set(r, c, v);
+	}
+
+	/**
+	 * Slice the current block and write into the outBlock. The offsets determines where the slice starts,
+	 * the length of the blocks is given by the outBlock dimensions.
+	 *
+	 * @param offsets  offsets where the slice starts
+	 * @param outBlock sliced result block
+	 * @return the sliced result block
+	 */
+	public TensorBlock slice(int[] offsets, TensorBlock outBlock) {
+		// TODO perf
+		int[] srcIx = offsets.clone();
+		int[] destIx = new int[offsets.length];
+		for (int l = 0; l < outBlock.getLength(); l++) {
+			outBlock.set(destIx, get(srcIx));
+			int i = outBlock.getNumDims() - 1;
+			destIx[i]++;
+			srcIx[i]++;
+			//calculating next index
+			while (destIx[i] == outBlock.getDim(i)) {
+				destIx[i] = 0;
+				srcIx[i] = offsets[i];
+				i--;
+				if (i < 0) {
+					//we are finished
+					return outBlock;
+				}
+				destIx[i]++;
+				srcIx[i]++;
+			}
+		}
+		return outBlock;
+	}
+
+	public TensorBlock copy(TensorBlock src) {
+		_dims = src._dims.clone();
+		_heterogeneous = src._heterogeneous;
+		if (_heterogeneous) {
+			_basicTensor = null;
+			_dataTensor = src._dataTensor == null ? null : new DataTensorBlock(src._dataTensor);
+		}
+		else {
+			_dataTensor = null;
+			_basicTensor = src._basicTensor == null ? null : new BasicTensorBlock(src._basicTensor);
+		}
+		return this;
+	}
+
+	public TensorBlock copy(int[] lower, int[] upper, TensorBlock src) {
+		if (_heterogeneous) {
+			if (src._heterogeneous) {
+				_dataTensor.copy(lower, upper, src._dataTensor);
+			}
+			else {
+				// TODO perf
+				_dataTensor.copy(lower, upper, new DataTensorBlock(src._basicTensor));
+			}
+		}
+		else {
+			if (src._heterogeneous) {
+				throw new DMLRuntimeException("Copying `DataTensor` into `BasicTensor` is not a safe operation.");
+			}
+			else {
+				_basicTensor.copy(lower, upper, src._basicTensor);
+			}
+		}
+		return this;
+	}
+
+	// `getExactSerializedSize()`, `write(DataOutput)` and `readFields(DataInput)` have to match in their serialized
+	// form definition
+	@Override
+	public long getExactSerializedSize() {
+		// header size (_isDataTensor, _dims.length + _dims[*], isAllocated)
+		long size = 1 + 4 * (1 + _dims.length) + 1;
+		if (isAllocated()) {
+			if (_heterogeneous) {
+				size += _dataTensor._schema.length;
+				for (BasicTensorBlock bt : _dataTensor._colsdata) {
+					if (bt != null)
```

Comment:
Why would we ever have a null basic tensor block in this list here?

Reply:
Because `_colsdata` will only contain `BasicTensorBlock`s for value types actually used in schema. I can rework this if necessary.